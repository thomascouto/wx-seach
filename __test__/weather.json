{
	"coord": {
		"lon": -87.65,
		"lat": 41.85
	},
	"weather": [
		{
			"id": 804,
			"main": "Clouds",
			"description": "overcast clouds",
			"icon": "04n"
		}
	],
	"base": "stations",
	"main": {
		"temp": 2.76,
		"feels_like": 2.76,
		"temp_min": 1.29,
		"temp_max": 3.63,
		"pressure": 1020,
		"humidity": 96
	},
	"visibility": 145,
	"wind": {
		"speed": 0.45,
		"deg": 51,
		"gust": 1.79
	},
	"clouds": {
		"all": 100
	},
	"dt": 1706187259,
	"sys": {
		"type": 2,
		"id": 2018264,
		"country": "US",
		"sunrise": 1706188185,
		"sunset": 1706223336
	},
	"timezone": -21600,
	"id": 4887398,
	"name": "Chicago",
	"cod": 200,
	"dt_txt": ""
}