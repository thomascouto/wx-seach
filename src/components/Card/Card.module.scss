.card {
  background-color: white;
  position: relative;
  display: grid;
  grid-template-columns: 1fr 1fr;
  align-items: center;
  gap: 0.5rem;
  width: 18rem;
  height: 11rem;
  filter: drop-shadow(-8px 5px 5px lightgray);

  img {
	position: absolute;
	left: -5rem;
	top: -1rem;
	width: 15rem;
  }

  div {
	grid-column-start: 2;
	display: flex;
	flex-direction: column;
	font-family: Lato, sans-serif;

	.temp {
		font-size: 3rem;
		font-weight: 400;

		&::after {
			content: 'Â°C';
			font-size: 2rem;
			position: absolute;
		}
	}

	span:nth-child(2), span:nth-child(3) {
		font-size: 0.7rem;
	}
  }
}
