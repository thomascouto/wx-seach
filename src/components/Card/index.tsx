import { ForecastCard } from '..';
import clsx from 'clsx';

import { useState } from 'react';

import { getIcon } from '@/util';

import styles from './Card.module.scss';

interface CardProps {
  data: WX;
  metricSystem: Units;
}

export default function Card({ data, metricSystem }: CardProps) {
  const [isExpanded, setIsExpanded] = useState(false);

  const {
    name,
    weather,
    wind,
    sys: { country },
    main: { humidity, temp },
  } = data;

  const handleForecast = () => {
    setIsExpanded((prev) => !prev);
  };

  const { main, description } = weather[0];

  return (
    <>
      <div className={styles.header}>
        {name}, {country}
        <button
          className={clsx(styles.forecast, isExpanded ? styles.hide : styles.expand)}
          onClick={handleForecast}
        />
      </div>
      <article className={styles.card}>
        <img src={'/wx/' + getIcon(main) + '.svg'} alt="Current weather" title={description} />
        <div className={styles.info}>
          <span className={styles.temp}>{Math.round(temp)}Â°</span>
          <span title="Humidity">
            ğŸ’§{humidity}% | {weather[0].main}
          </span>
          <span title="Wind">{`ğŸ’¨ ${wind.deg.toString().padStart(3, '0')}Â° at ${Math.round(wind.speed)}`}</span>
        </div>
      </article>
      {isExpanded && <ForecastCard location={name} units={metricSystem} />}
    </>
  );
}
