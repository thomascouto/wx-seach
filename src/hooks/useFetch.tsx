import { useState } from 'react';

import { getWeatherService } from '@/services';

export default function useFetch<T>() {
  const [weather, setWeather] = useState<T>();
  const [error, setError] = useState<ErrorResponse>();
  const [isLoading, setIsLoading] = useState(false);

  const getWeatherData = ({ q, isForecast, units = 'metric' }: QueryProps) => {
    setIsLoading(true);
    getWeatherService<T>({
      q,
      units,
      isForecast,
    })
      .then(setWeather)
      .catch((err: string) => {
        const parsedError = JSON.parse(err) as ErrorResponse;
        setError(parsedError);
      })
      .finally(() => {
        setIsLoading(false);
      });
  };

  return {
    getWeatherData,
    weather,
    isLoading,
    error,
  };
}
