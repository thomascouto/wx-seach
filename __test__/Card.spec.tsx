import expandedWeather from './expanded.json';
import weather from './weather.json';

import { fireEvent, render, screen } from '@testing-library/react';

import Card from '@/components/Card';
import * as useFetch from '@/hooks/useFetch';

describe('Card tests', () => {
  it('should render correctly', () => {
    const { container } = render(<Card data={weather} metricSystem="metric" />);

    expect(container).toMatchSnapshot();
  });

  it('should expand card', () => {
    jest.spyOn(useFetch, 'default').mockImplementation(() => ({
      getWeatherData: jest.fn(),
      weather: expandedWeather,
      isLoading: false,
      error: null,
    }));

    const { container } = render(<Card data={weather} metricSystem="metric" />);

    const btn = screen.getByRole('button');

    fireEvent.click(btn);

    expect(btn.classList.toString()).toBe('forecast hide');
    expect(container).toMatchSnapshot();
  });
});
