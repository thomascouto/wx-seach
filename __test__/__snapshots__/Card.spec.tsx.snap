// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card tests should expand card 1`] = `
<div>
  <div
    class="header"
  >
    Chicago
    , 
    US
    <button
      class="forecast hide"
    />
  </div>
  <article
    class="card"
  >
    <img
      alt="Current weather"
      src="/wx/cloudy-day.svg"
      title="overcast clouds"
    />
    <div
      class="info"
    >
      <span
        class="temp"
      >
        3
        Â°
      </span>
      <span
        title="Humidity"
      >
        ðŸ’§
        96
        % | 
        Clouds
      </span>
      <span
        title="Wind"
      >
        ðŸ’¨ 051Â° at 0
      </span>
    </div>
  </article>
  <article
    class="forecastWeather"
  >
    <div
      class="forecastData"
    >
      <span>
        2024-01-26
      </span>
      <img
        alt="Wind 48Â° at 1.92"
        src="/arrow.png"
        style="width: 15px;"
      />
      <span>
        <img
          src="/hot-temp.png"
          width="20"
        />
        1
        Â°
      </span>
      <span>
        <img
          src="/low-temp.png"
          width="20"
        />
        1
        Â°
      </span>
      <span>
        ðŸ’§
        100
        %
      </span>
      <img
        alt="Weather"
        src="/wx/rainy.svg"
        width="160px"
      />
      <span
        class="temperature"
      >
        1
        Â°
      </span>
    </div>
    <div
      class="forecastData"
    >
      <span>
        2024-01-27
      </span>
      <img
        alt="Wind 334Â° at 1.27"
        src="/arrow.png"
        style="width: 15px;"
      />
      <span>
        <img
          src="/hot-temp.png"
          width="20"
        />
        1
        Â°
      </span>
      <span>
        <img
          src="/low-temp.png"
          width="20"
        />
        1
        Â°
      </span>
      <span>
        ðŸ’§
        99
        %
      </span>
      <img
        alt="Weather"
        src="/wx/rainy.svg"
        width="160px"
      />
      <span
        class="temperature"
      >
        1
        Â°
      </span>
    </div>
    <div
      class="forecastData"
    >
      <span>
        2024-01-28
      </span>
      <img
        alt="Wind 16Â° at 2.53"
        src="/arrow.png"
        style="width: 15px;"
      />
      <span>
        <img
          src="/hot-temp.png"
          width="20"
        />
        2
        Â°
      </span>
      <span>
        <img
          src="/low-temp.png"
          width="20"
        />
        2
        Â°
      </span>
      <span>
        ðŸ’§
        99
        %
      </span>
      <img
        alt="Weather"
        src="/wx/cloudy-day.svg"
        width="160px"
      />
      <span
        class="temperature"
      >
        2
        Â°
      </span>
    </div>
    <div
      class="forecastData"
    >
      <span>
        2024-01-29
      </span>
      <img
        alt="Wind 342Â° at 2.03"
        src="/arrow.png"
        style="width: 15px;"
      />
      <span>
        <img
          src="/hot-temp.png"
          width="20"
        />
        2
        Â°
      </span>
      <span>
        <img
          src="/low-temp.png"
          width="20"
        />
        2
        Â°
      </span>
      <span>
        ðŸ’§
        88
        %
      </span>
      <img
        alt="Weather"
        src="/wx/cloudy-day.svg"
        width="160px"
      />
      <span
        class="temperature"
      >
        2
        Â°
      </span>
    </div>
    <div
      class="forecastData"
    >
      <span>
        2024-01-30
      </span>
      <img
        alt="Wind 198Â° at 3.44"
        src="/arrow.png"
        style="width: 15px;"
      />
      <span>
        <img
          src="/hot-temp.png"
          width="20"
        />
        2
        Â°
      </span>
      <span>
        <img
          src="/low-temp.png"
          width="20"
        />
        2
        Â°
      </span>
      <span>
        ðŸ’§
        79
        %
      </span>
      <img
        alt="Weather"
        src="/wx/cloudy-day.svg"
        width="160px"
      />
      <span
        class="temperature"
      >
        2
        Â°
      </span>
    </div>
  </article>
</div>
`;

exports[`Card tests should render correctly 1`] = `
<div>
  <div
    class="header"
  >
    Chicago
    , 
    US
    <button
      class="forecast expand"
    />
  </div>
  <article
    class="card"
  >
    <img
      alt="Current weather"
      src="/wx/cloudy-day.svg"
      title="overcast clouds"
    />
    <div
      class="info"
    >
      <span
        class="temp"
      >
        3
        Â°
      </span>
      <span
        title="Humidity"
      >
        ðŸ’§
        96
        % | 
        Clouds
      </span>
      <span
        title="Wind"
      >
        ðŸ’¨ 051Â° at 0
      </span>
    </div>
  </article>
</div>
`;
